src/colvarbias_opes.cpp

```diff
@@ -0,0 +1,1977 @@
+    }
+  }
+  if (m_no_zed) {
+    printInfo("noZed: using fixed normalization factor = ", cvm::to_str(m_zed));
+  }
+  if (comm == multiple_replicas && m_num_walkers > 1) {
+    cvm::log(this->name + " if multiple replicas are present, they will share the same bias\n");
+  }
+  if (m_num_threads > 1) {
+    printInfo("using multiple threads per simulation: ", cvm::to_str(m_num_threads));
+  }
+  cvm::main()->cite_feature("OPES");
+  if (m_adaptive_sigma || m_explore) {
+    cvm::main()->cite_feature("OPES explore or adaptive kernels");
+  }
+}
+
+cvm::real colvarbias_opes::evaluateKernel(
+  const colvarbias_opes::kernel& G,
+  const std::vector<cvm::real>& x) const {
+  cvm::real norm2 = 0;
+  for (size_t i = 0; i < num_variables(); ++i) {
+    const cvm::real dist2_i = variables(i)->dist2(G.m_center[i], x[i]) / (G.m_sigma[i] * G.m_sigma[i]);
+    norm2 += dist2_i;
+    if (norm2 >= m_cutoff2) {
+      return 0;
+    }
+  }
+  return G.m_height * (std::exp(-0.5 * norm2) - m_val_at_cutoff);
+}
+
+cvm::real colvarbias_opes::evaluateKernel(
+  const colvarbias_opes::kernel& G,
+  const std::vector<cvm::real>& x,
+  std::vector<cvm::real>& accumulated_derivative,
+  std::vector<cvm::real>& dist) const {
+  cvm::real norm2 = 0;
+  for (size_t i = 0; i < num_variables(); ++i) {
+    dist[i] = 0.5 * variables(i)->dist2_lgrad(x[i], G.m_center[i]) / G.m_sigma[i];
+    norm2 += dist[i] * dist[i];
+    if (norm2 >= m_cutoff2) {
+      return 0;
+    }
+  }
+  const cvm::real val = G.m_height * (std::exp(-0.5 * norm2) - m_val_at_cutoff);
+  // The derivative of norm2 with respect to x
+  for (size_t i = 0; i < num_variables(); ++i) {
+    accumulated_derivative[i] -= val * dist[i] / G.m_sigma[i];
+  }
+  return val;
+}
+
+cvm::real colvarbias_opes::getProbAndDerivatives(
+  const std::vector<cvm::real>& cv, std::vector<cvm::real>& der_prob) const {
+  double prob = 0.0;
+  std::vector<cvm::real> dist(num_variables(), 0);
+  if (!m_nlist) {
+    if (m_num_threads == 1 || m_kernels.size() < 2 * m_num_threads) {
+      for (size_t k = 0; k < m_kernels.size(); ++k) {
+        prob += evaluateKernel(m_kernels[k], cv, der_prob, dist);
+      }
+    } else {
+#if defined(_OPENMP)
+      #pragma omp parallel num_threads(m_num_threads)
+      {
+        std::vector<cvm::real> omp_deriv(der_prob.size(), 0);
+        std::vector<cvm::real> tmp_dist(num_variables());
+        #pragma omp for reduction(+:prob) nowait
+        for (size_t k = 0; k < m_kernels.size(); ++k) {
+          prob += evaluateKernel(m_kernels[k], cv, omp_deriv, tmp_dist);
+        }
+        #pragma omp critical
+        for (size_t i = 0; i < num_variables(); ++i) {
+          der_prob[i]+=omp_deriv[i];
+        }
+        #pragma omp single
+        for (size_t i = 0; i < num_variables(); ++i) {
+          dist[i] = tmp_dist[i];
+        }
+      }
+#elif defined(CMK_SMP) && defined(USE_CKLOOP)
+      // TODO: Does this work??
```

Comment:
Is this comment still current?

Reply:
Yes. It is still on my to-do list. Once Colvars switches to support a fine-grained parallelization scheme, I will test these branches.